nodes/explorer/src/api_backend/store.rs

```diff
@@ -44,3 +46,184 @@ where
         .collect();
     make_request_and_return_response!(futures::future::try_join_all(results))
 }
+
+type Depth = usize;
+
+#[derive(Deserialize)]
+pub(crate) struct BlocksByIdQueryParams {
+    from: BlockId,
+    to: Option<BlockId>,
+}
+
+pub(crate) async fn blocks<Tx, S>(
+    State(store): State<OverwatchHandle>,
+    Query(query): Query<BlocksByIdQueryParams>,
+) -> Response
+where
+    Tx: Transaction
+        + Clone
+        + Debug
+        + Eq
+        + Hash
+        + Serialize
+        + DeserializeOwned
+        + Send
+        + Sync
+        + 'static,
+    <Tx as Transaction>::Hash: std::cmp::Ord + Debug + Send + Sync + 'static,
+    S: StorageSerde + Send + Sync + 'static,
+{
+    let BlocksByIdQueryParams { from, to } = query;
+    // get the from block
+    let from = match storage::block_req::<S, Tx>(&store, from).await {
+        Ok(from) => match from {
+            Some(from) => from,
+            None => {
+                return IntoResponse::into_response((StatusCode::NOT_FOUND, "from block not found"))
+            }
+        },
+        Err(e) => {
+            return IntoResponse::into_response((StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))
+        }
+    };
+
+    // check if to is valid
+    match to {
+        Some(to) => match storage::block_req::<S, Tx>(&store, to).await {
+            Ok(to) => match to {
+                Some(to) => handle_to::<S, Tx>(store, from, Some(to)).await,
+                None => IntoResponse::into_response((StatusCode::NOT_FOUND, "to block not found")),
+            },
+            Err(e) => {
+                IntoResponse::into_response((StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))
+            }
+        },
+        None => handle_to::<S, Tx>(store, from, None).await,
+    }
+}
+
+async fn handle_to<S, Tx>(
+    store: OverwatchHandle,
+    from: Block<Tx, Certificate>,
+    to: Option<Block<Tx, Certificate>>,
+) -> Response
+where
+    Tx: Transaction
+        + Clone
+        + Debug
+        + Eq
+        + Hash
+        + Serialize
+        + DeserializeOwned
+        + Send
+        + Sync
+        + 'static,
+    <Tx as Transaction>::Hash: std::cmp::Ord + Debug + Send + Sync + 'static,
+    S: StorageSerde + Send + Sync + 'static,
+{
+    let mut current = Some(from.header().parent());
+    let mut blocks = Vec::new();
+    while let Some(id) = current {
```

Comment:
What do you think if storage service would be responsible for the aggregation of this data? If we had a dedicated storage method for getting a range of blocks, then this logic could be contained there, and http service would only manage routing with params?

Reply:
That would depend on the storage backend actually. And it may change. I think for now this is ok. It can be moved to an specific storage module itself, ot even try in a single query. But definitely the storage service is abstracted from the usage of it.

Reply:
Yes, ideally, this should be done in the storage service, but it will require breaking changes on the current storage services.

Reply:
https://www.notion.so/Block-Explorer-4d32e6f08c144a3a9166d25512e0e339?pvs=4#15330f508e944d289acb5f6cb2f7a4fe This explains why we cannot do such blocks aggregation in the storage layer. 

And here is an example, how `sled` works:

```rust
#[test]
    fn test_sled() {
        let db = sled::Config::default()
            .temporary(true)
            .open()
            .expect("failed to open sled db");

        db.insert(1usize.to_be_bytes(), 1usize.to_be_bytes().as_ref());
        db.insert(2usize.to_be_bytes(), 2usize.to_be_bytes().as_ref());
        db.insert(6usize.to_be_bytes(), 6usize.to_be_bytes().as_ref());
        db.insert(5usize.to_be_bytes(), 5usize.to_be_bytes().as_ref());

        let (last_k, _) = db.last().unwrap().unwrap();
        assert_eq!(last_k, 5usize.to_be_bytes()); // panic: last_k is 6, because of sled is sorted by key
    }
```

Reply:
For now this naive solution should be enough. We do not even know if we will keep the storage as it is. We may chose something different in the future as we are changing a lot of stuff. We just need a replacement to reduce consensus service  responsibilities.